---
title: "PHATE analysis CuratedMetaGData version 3"
output: html_notebook
---





```{r}

library(phyloseq)
devtools::load_all()
library(magrittr)
library(forcats)
library(ggplot2)
library(tidygraph)
library(dplyr)

```



```{r}

load(system.file("data-raw/curatedMetaG/curated_v3_otu_tax.rda", package = "gutzilla"))

enterotypes = read.csv2("enterotypes_curated_v3_prediction.csv", row.names=1)


```



```{r}

OTU %>%
  merge(TAX %>% as.data.frame() %>% select(Genus), by="row.names") %>%
  select(-Row.names) %>%
  group_by(Genus) %>%
  summarise_all(sum) -> curated_v3_genus
  
  

curated_v3_genus %>%
  arrange(desc(SAMEA7041268))


curated_v3_genus$Genus

grep("Muribaculaceae",row.names(TAX), value = TRUE)



```

```{r}


OTU %>%
  merge(TAX %>% as.data.frame() %>% select(Species), by="row.names") %>%
  select(-Row.names) %>%
  group_by(Species) %>%
  summarise_all(sum) -> curated_v3_species

curated_v3_genus %>%
  ungroup() %>%
  mutate_if(is.numeric, function(x) x/sum(x)) %>%
  tibble::column_to_rownames("Genus") %>% 
  as.matrix %>% 
  apply(1,sum) %>% 
  sort %>% rev %>% head(100) -> top_species_mass_curated


curated_v3_species %>%
  ungroup() %>%
  mutate_if(is.numeric, function(x) x/sum(x)) -> curated_v3_species_prop



save(curated_v3_species, file="curated_v3_species_count.rda")

save(curated_v3_species_prop, file="curated_v3_species_prop.rda")

```


```{r}


curated_v3_genus %>%
  ungroup() %>%
  mutate_if(is.numeric, function(x) x/sum(x)) %>%
  tibble::column_to_rownames("Genus") %>% 
  as.matrix %>% 
  apply(1,sum) %>% 
  sort %>% rev %>% head(30) -> top_genus_mass_curated



curated_v3_genus %>%
  ungroup() %>%
  mutate_if(is.numeric, function(x) x/sum(x)) -> curated_v3_genus_prop


curated_v3_genus %>%
  filter(Genus %in% names(top_genus_mass_curated)) %>%
  summarise_if(is.numeric, sum) -> dominant_mass_count_per_sample


curated_v3_genus_prop %>%
  filter(Genus %in% names(top_genus_mass_curated)) %>%
  summarise_if(is.numeric, sum) -> dominant_mass_per_sample



dominant_mass_per_sample %>%
  t %>%
  as.data.frame() %>%
  arrange(V1) %>%
  filter(V1>0.25) %>% row.names() -> sample_curated_to_select


curated_v3_genus_prop %>%
  filter(Genus %in% names(top_genus_mass_curated)) %>% 
  tibble::column_to_rownames("Genus") %>% as.matrix %>% cor() -> sample_cor


curated_v3_genus_prop %>%
  mutate_if(is.numeric, function(x) round(x*10^4,0)  ) %>%
  filter(Genus %in% names(top_genus_mass_curated)) %>%
  tibble::column_to_rownames("Genus") -> genus_count_dominant


```
## alpha and beta between train and remaining set

```{r}

curated_v3_genus_prop %>%
     filter(Genus %in% c(names(top_genus_mass_curated),"Methanobrevibacter")) %>%
     tibble::column_to_rownames("Genus") %>%
     select(all_of(sample_curated_to_select)) %>% t %>% Rfast::Dist(method = "jensen_shannon") -> curated_v3_genus_prop_jsd


curated_v3_genus_prop_jsd_sqrt = sqrt(curated_v3_genus_prop_jsd)



rownames(curated_v3_genus_prop_jsd_sqrt) = colnames(curated_v3_genus_prop_jsd_sqrt) = 
  curated_v3_genus_prop %>%
     filter(Genus %in% c(names(top_genus_mass_curated),"Methanobrevibacter")) %>%
     tibble::column_to_rownames("Genus") %>%
     select(all_of(sample_curated_to_select)) %>% colnames()



curated_v3_genus_prop_jsd_sqrt

vegan::adonis()


vegan::adonis(as.dist(curated_v3_genus_prop_jsd_sqrt[1:100,1:100])~d, data.frame(d=rep(c("A","B"),50))) %>% .$aov.tab %>% broom::tidy() %>% .[1,6:7]

adonis_R2_p = NULL

for(i in 1:24) {

  #enterotypes %>% filter(Enterotypes_id==1) %>% pull(sample_name)
  
vegan::adonis(
  
  as.dist(
    
    curated_v3_genus_prop_jsd_sqrt[
      enterotypes %>% filter(Enterotypes_id==i) %>% pull(sample_name),
      enterotypes %>% filter(Enterotypes_id==i) %>% pull(sample_name)])~set, 
  
  enterotypes %>% filter(Enterotypes_id==i)
  ) %>% 
  .$aov.tab %>% 
  broom::tidy() %>% 
  .[1,6:7] -> tmp

adonis_R2_p = rbind(adonis_R2_p,tmp)

print(i)
}



adonis_R2_p %>% mutate(p.value=p.adjust(p.value, method="fdr"))



enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  group_by(Enterotypes_id) %>%
  do(wilcox.test(.$shannon~.$set, data=.)%>%broom::tidy(), ) %>%
  mutate(p.value=p.adjust(p.value,method="fdr")) %>%
  as.data.frame %>%
  cbind(data.frame(adonis_R2_p)) %>%
  .[,c(-4,-5)] %>%
  mutate(p.value.1=p.adjust(p.value.1, method="fdr")) %>%
  dplyr::rename(microbiome_partition=Enterotypes_id, alpha.p.value = p.value, beta.p.value=p.value.1,wilcoxon_statistic=statistic, adonis_R2=R2) %>%
  readr::write_csv2(file="figures/table_alpha_beta_train_remaining.csv")


enterotypes %>% select(Enterotypes_id,set) %>% ggplot + geom_bar(aes(x=Enterotypes_id,fill=set), position = "fill") + geom_hline(yintercept = 0.2)


```


## compute PHATE plot
```{r}

genus_phate = 
  curated_v3_genus_prop %>%
  filter(Genus %in% c(names(top_genus_mass_curated),"Methanobrevibacter")) %>%
  tibble::column_to_rownames("Genus") %>%
  select(all_of(sample_curated_to_select)) %>%
  t() %>%
  #as.data.frame() %>%
  #merge(enterotypes %>% filter(set != "outliers", value>0.80) %>% select(sample_name), by.x="row.names", by.y="sample_name") %>% head
  
  phateR::phate(gamma=0, t=40, seed=666)


# genus_phate_test = 
#   curated_v3_genus_prop %>%
#   filter(Genus %in% c(names(top_genus_mass_curated),"Methanobrevibacter")) %>%
#   tibble::column_to_rownames("Genus") %>%
#   select(all_of(sample_curated_to_select)) %>%
#   t() %>%
#   #as.data.frame() %>%
#   #merge(enterotypes %>% filter(set != "outliers", value>0.80) %>% select(sample_name), by.x="row.names", by.y="sample_name") %>% head
#   
#   phateR::phate(gamma=1, seed=666, decay=40, t=40)


curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=P_B_ratio %>% log10())) + scale_color_viridis_c("log10(P/B ratio)")


```

### PHATE 3D

```{r}


genus_phate_3D = 
  curated_v3_genus_prop %>%
  filter(Genus %in% c(names(top_genus_mass_curated),"Methanobrevibacter")) %>%
  tibble::column_to_rownames("Genus") %>%
  select(all_of(sample_curated_to_select)) %>%
  t() %>%
  #as.data.frame() %>%
  #merge(enterotypes %>% filter(set != "outliers", value>0.80) %>% select(sample_name), by.x="row.names", by.y="sample_name") %>% head
  
  phateR::phate(gamma=0, seed=666, ndim=3)

save(genus_phate_3D, curated_v3_genus, file="phate_3D_genus_count.rda")

curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate_3D$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=P_B_ratio %>% log10())) + scale_color_viridis_c("log10(P/B ratio)")


curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate_3D$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE3, col=P_B_ratio %>% log10())) + scale_color_viridis_c("log10(P/B ratio)")



curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate_3D$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>% 
plotly::plot_ly(x=.[,"PHATE1"], y=.[,"PHATE2"], z=.[,"PHATE3"], type="scatter3d", color=.$P_B_ratio %>% log10())

curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate_3D$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>% 
plotly::plot_ly(x=.[,"PHATE1"], y=.[,"PHATE2"], z=.[,"PHATE3"], type="scatter3d", color=.$P_B_ratio %>% log10(), size = 1)





```



```{r}


curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Prevotella") %>% 
  merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(P_B_ratio = (`Prevotella`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=P_B_ratio %>% log2()), size=1, alpha=0.5) + scale_color_viridis_c("Prevotella/\nBacteroides\nratio (log2)", option="cividis") + cowplot::theme_cowplot() -> fig1C


# curated_v3_genus %>%
#   tibble::column_to_rownames("Genus") %>%
#   t() %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column("sample_name") %>%
#   select("sample_name", "Bacteroides","Akkermansia") %>% 
#   merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
#   mutate(A_B_ratio = (`Akkermansia`+1)/(`Bacteroides`+1)) %>%
#   ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=A_B_ratio %>% log2())) + scale_color_viridis_c("log10(A/B ratio)")
# 
# 
# curated_v3_genus %>%
#   tibble::column_to_rownames("Genus") %>%
#   t() %>%
#   as.data.frame() %>%
#   tibble::rownames_to_column("sample_name") %>%
#   select("sample_name", "Bacteroides","Methanobrevibacter") %>% 
#   merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
#   mutate(M_B_ratio = (`Methanobrevibacter`+1)/(`Bacteroides`+1)) %>%
#   ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=M_B_ratio %>% log2())) + scale_color_viridis_c("log10(M/B ratio)")


curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Bifidobacterium") %>% 
  merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(B_B_ratio = (`Bifidobacterium`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=B_B_ratio %>% log2()), size=1, alpha=0.5) + scale_color_viridis_c("Bifidobacterium/\nBacteroides\nratio (log2)", option="cividis") + cowplot::theme_cowplot() -> fig1D


curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select("sample_name", "Bacteroides","Escherichia") %>% 
  merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") %>%
  mutate(E_B_ratio = (`Escherichia`+1)/(`Bacteroides`+1)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y= PHATE2, col=E_B_ratio %>% log2()), size=1, alpha=0.5) + scale_color_viridis_c("Escherichia/\nBacteroides\nratio (log2)", option="cividis") + cowplot::theme_cowplot()



```


## load DMM models



```{r fig.height=6, fig.width=10}

load("genus_alpha_weight_curated.rda")

genus_alpha_weight_curated %>%
  
ggplot() + 
  geom_tile(aes(x=variable,y=taxa,fill=log10(value))) + 
  scale_fill_viridis_c("DMM model\nalpha (log10)") +
  ylab("taxa") +
  xlab("DMM community types") -> figS5A


enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
  group_by(Enterotypes_id) %>%
  summarise(shannon=median(shannon)) %>%
  merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
  ggplot() + 
  geom_tile(aes(x=Enterotypes_id,y=taxa,fill=log10(value))) + 
  scale_fill_viridis_c("DMM alpha\nweight (log10)") +
  ylab("taxa") +
  xlab("DMM community types") +
  cowplot::theme_cowplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) -> figS5A



fig1B = figS5A




```



```{r}

enterotypes = readr::read_csv2("enterotypes_curated_v3_prediction.csv")[,-1]

genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% as.character) %>%
  group_by(Enterotypes_id) %>%
  summarise(PHATE1=mean(PHATE1), PHATE2=mean(PHATE2)) %>%
  mutate(peripherality= abs(PHATE1) + abs(PHATE2)) -> enterotypes_phate_centroid

enterotypes_phate_centroid %>%
  arrange(PHATE1)

enterotypes_phate_centroid %>%
  arrange(PHATE2)


enterotypes_phate_centroid %>%
  arrange(peripherality)
  
  

enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
  group_by(Enterotypes_id) %>%
  summarise(shannon=median(shannon)) %>%
  merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
  mutate(taxa2 = ifelse(taxa %in% c("Faecalibacterium","Ruminococcus","Blautia","Roseburia","Lachnospiraceae unclassified","Oscillibacter","Ruminococcaceae unclassified","Eubacterium") , "Clostridiales", as.character(taxa))) %>%
  group_by(Enterotypes_id, taxa2, shannon) %>%
  summarise(value=sum(value)) %>%
  ungroup() %>%
  group_by(Enterotypes_id) %>%
  top_n(n = 1, value) %>%
  arrange(shannon)




genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  mutate(branch ="Clostridiales") %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(16,12,4,13,17,14), "Prevotella",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(21,20,5,3,2), "Bacteroides",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(23,24), "Bifidobacterium",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(19,22), "AeroAnaerobic",branch)) %>%
  #mutate(branch = ifelse(Enterotypes_id %in% c(7,11,12), "Akkermansia DMM types",branch)) %>%
  #filter(!(Enterotypes_id %in% c(19,20,21,23,24))) %>% # remove low theta dmm
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=branch), size=1, alpha=0.5) +
  scale_color_brewer("main taxa\nDMM types",type="qual") + theme_classic() +
  ggrepel::geom_label_repel(data = enterotypes_phate_centroid, aes(x=PHATE1, y=PHATE2, label=Enterotypes_id), max.overlaps = 20) -> fig1H

fig1H

```

```{r fig.height=6, fig.width=12}


genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  mutate(branch ="Clostridiales") %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(16,12,4,13,17,14), "Prevotella",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(21,20,5,3,2), "Bacteroides",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(23,24), "Bifidobacterium",branch)) %>%
  mutate(branch = ifelse(Enterotypes_id %in% c(19,22), "AeroAnaerobic",branch)) %>%
  #mutate(branch = ifelse(Enterotypes_id %in% c(7,11,12), "Akkermansia DMM types",branch)) %>%
  #filter(!(Enterotypes_id %in% c(19,20,21,23,24))) %>% # remove low theta dmm
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=branch), size=1, alpha=0.5) +
  scale_color_brewer("main taxa\nDMM types",type="qual") + theme_classic() + facet_wrap(~branch)
  

```




```{r}

MN_species = readLines(system.file("data-raw/MN_species.txt", package = "gutzilla"))
MH_species = readLines(system.file("data-raw/MH_species.txt", package = "gutzilla"))

MN_species_tax = grep(paste(MN_species,sep="|", collapse = "|"),row.names(OTU), value=TRUE  )
MH_species_tax = grep(paste(MH_species,sep="|", collapse = "|"),row.names(OTU), value=TRUE  )






```

```{r}

MH_species_metagenome =
OTU %>% 
  unclass %>%
  prop.table(2) %>%
  as.data.frame %>%
  tibble::rownames_to_column("taxa") %>%
  filter(taxa %in% c(MH_species_tax)) %>%
  tibble::column_to_rownames("taxa")

MN_species_metagenome =
OTU %>% 
  unclass %>%
  prop.table(2) %>%
  as.data.frame %>%
  tibble::rownames_to_column("taxa") %>%
  filter(taxa %in% c(MN_species_tax)) %>%
  tibble::column_to_rownames("taxa")
  


```





## GMHI computation

```{r}


output_file = 'GMHI_curated_v3_output.csv'

alpha <- function(x){sum((log(x[x>0]))*(x[x>0]))*(-1)}

MH_shannon <- apply((MH_species_metagenome), 2, alpha) 

MN_shannon <- apply((MN_species_metagenome), 2, alpha) 





# Richness of Health-prevalent species

# Richness of Health-scarce species

R_MH <- apply(MH_species_metagenome, 2, function(i) (sum(i > 0))) 

R_MN <- apply(MN_species_metagenome, 2, function(i) (sum(i > 0)))





# Median RMH from 1% of the top-ranked samples

# Median RMN from 1% of the bottom-ranked samples

# Supplementary Methods for further details

MH_prime <- 7

MN_prime <- 31





# Collective abundance of Health-prevalent species

# Collective abundance of Health-scarce species

psi_MH <- ((R_MH/MH_prime)*MH_shannon) 

psi_MN <- ((R_MN/MN_prime)*MN_shannon)



GMHI <- data.frame(log10((psi_MH+0.00001)/(psi_MN+0.00001))) # 0.00001 added to avoid having the denominator as 0

colnames(GMHI) <- c("GMHI")



if (file.exists(output_file)){

	file.remove(output_file)

}

write.csv(GMHI, file=output_file) # Saving GMHI results as 'GMHI_curated_v3_output.csv'. User should change the path to the appropriate directory.




```


```{r}


genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  merge(GMHI, by.x="Row.names", by.y="row.names" ) %>%
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=GMHI), size=1, alpha=0.5) +
  scale_color_viridis_c("GMHI", option="cividis") + cowplot::theme_cowplot() -> fig1F_gmhi
#+
  #ggrepel::geom_label_repel(data = enterotypes_phate_centroid, aes(x=PHATE1, y=PHATE2, label=Enterotypes_id)) 

fig1F_gmhi

ggsave("gmhi_curated_v3_phate.pdf")

```

# Alpha-diversity
```{r}

shannon_curated_v3 =
OTU %>%
  unclass %>%
  as.data.frame() %>%
  t() %>% 
  vegan::rrarefy(sample=10^6) %>%
  vegan::diversity(MARGIN = 1) %>%
  as.data.frame() %>%
  dplyr::rename(shannon = 1)

save(shannon_curated_v3, file="shannon_curated_v3.rda")

```



```{r}

genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  merge(shannon_curated_v3, by.x="Row.names", by.y="row.names" ) %>%
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=shannon), size=1, alpha=0.5) +
  scale_color_viridis_c("Shannon", option="cividis") + 
  cowplot::theme_cowplot() -> fig1F #+
  #ggrepel::geom_label_repel(data = enterotypes_phate_centroid, aes(x=PHATE1, y=PHATE2, label=Enterotypes_id)) -> fig1F

fig1F

ggsave("shannon_curated_v3_phate.pdf")


```

## dominance


```{r}


OTU %>% #.[1:500,1:500] %>%
  prop.table(2) %>%
  .^2 %>%
  apply(2,sum) %>%
  data.frame() %>% 
  dplyr::rename(dominance=1) -> dominance

genus_phate$embedding %>% 
  #merge(enterotypes, by.x="row.names", by.y="sample_name") %>% head
  merge(dominance, by.x="row.names", by.y="row.names" ) %>% 
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=dominance), size=1, alpha=0.5) +
  scale_color_viridis_c("dominance", option="cividis", direction = -1) + cowplot::theme_cowplot() #+
  #ggrepel::geom_label_repel(data = enterotypes_phate_centroid, aes(x=PHATE1, y=PHATE2, label=Enterotypes_id))

```




## predict lineage with slingshot



```{r}

library(slingshot)


enterotypes_curated_v3 = read.csv2("enterotypes_curated_v3_prediction.csv", row.names = 1)


genus_phate_enterotypes_curated_v3 = merge(genus_phate$embedding,enterotypes_curated_v3, by.x="row.names", by.y="sample_name")


genus_phate_enterotypes_curated_v3 %>%
  dplyr::sample_n(size=1000) -> genus_phate_enterotypes_curated_v3_sub
      
      
# pto <- slingshot(
#   genus_phate_enterotypes_curated_v3_sub %>% select(PHATE1,PHATE2),
#   genus_phate_enterotypes_curated_v3_sub$Enterotypes_id , approx_points = 150)
# 
# 
# lin1 <- getLineages(genus_phate_enterotypes_curated_v3_sub %>% select(PHATE1,PHATE2),
#   genus_phate_enterotypes_curated_v3_sub$Enterotypes_id %>% as.character() )
# 
# lin1
# 
# plot(genus_phate_enterotypes_curated_v3_sub %>% select(PHATE1,PHATE2),  asp = 1)
# plot(lin1, lwd = 3, col = 'black')
# 
# 
# pseudo <- slingPseudotime(pto)
# curve_weight <- slingCurveWeights(pto)      
# 
# 
# 
# genus_phate_enterotypes_curated_v3_sub
# 
# plot(genus_phate_enterotypes_curated_v3_sub %>% select(PHATE1,PHATE2),  asp = 1)
# lines(pto, lwd = 3)




genus_phate_enterotypes_curated_v3 %>%
  group_by(Enterotypes_id) %>%
  summarise(PHATE1=mean(PHATE1),PHATE2=mean(PHATE2)) %>%
  tibble::column_to_rownames("Enterotypes_id") %>%
  dist() %>%
  ape::mst() -> M


genus_phate_enterotypes_curated_v3 %>% 
  #filter(set=="train") %>%
  
  ggplot() + geom_bar(aes(x=Enterotypes_id, fill=Enterotypes_id)) + facet_wrap(~set)


phate_dmm_centroid = genus_phate_enterotypes_curated_v3 %>%
  #filter(set=="train") %>%
  group_by(Enterotypes_id) %>%
  summarise(x=mean(PHATE1),y=mean(PHATE2), n=n()) %>%
  tibble::column_to_rownames("Enterotypes_id")

plot(M, x1 = phate_dmm_centroid$PHATE1, x2 = phate_dmm_centroid$PHATE2)

M %>%  
  igraph::graph_from_adjacency_matrix() %>%
  as_tbl_graph() %>%
  tidygraph::activate(nodes) %>%
  mutate(size =phate_dmm_centroid$n ) %>%
  ggraph(layout = 'kk') + 
    geom_edge_link() + 
    geom_node_point(mapping=aes(size=log(size^2)) , colour = 'steelblue') +
    geom_node_text(aes(label = name), colour = 'white', vjust = 0.4) 





M %>%  
  igraph::graph_from_adjacency_matrix() %>%
  ggraph(layout = 'manual', circular= FALSE, x = phate_dmm_centroid[,1], y=phate_dmm_centroid[,2]) + 
  geom_point(data = as.data.frame(genus_phate$embedding), aes(x=PHATE1, y=PHATE2), color="grey", alpha=0.1, size=1) +
    geom_edge_link() + 
    geom_node_point(size=6, data = phate_dmm_centroid, mapping=aes(color=n) ) + 
  scale_color_gradient("Nb of\nsamples") +
    geom_node_text(aes(label = name), colour = 'white', vjust = 0.4) + 
  cowplot::theme_cowplot() + xlab("PHATE 1") + ylab("PHATE 2") -> fig1G

fig1G



# phate_dmm_centroid %>%
#   tibble::rownames_to_column("partition") %>%
#   ggplot() + geom_bar(aes(y=n, x=partition %>% as.character), stat=identity())
      
```




```{r}

shannon_curated_v3


enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
  group_by(Enterotypes_id) %>%
  summarise(shannon=median(shannon)) %>%
  merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
  group_by(Enterotypes_id) %>%
  summarise(shannon = median(shannon)) %>%
  mutate(alpha = "alpha") %>%
ggplot() + 
  geom_tile(aes(x=Enterotypes_id, y=alpha,fill=shannon)) + 
  scale_fill_continuous("alpha-diversity median") +
  ylab("") +
  xlab("DMM community types")


enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
  merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
  ggplot() + 
  geom_boxplot(aes(x=Enterotypes_id, y=shannon)) + 
  ylab("alpha-diversity") +
  xlab("DMM community types") +
  cowplot::theme_cowplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) -> fig1A

# enterotypes %>%
#   merge(dominance, by.x="sample_name", by.y="row.names" ) %>%
#   mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
#   merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
#   mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(dominance, median, .desc = TRUE)) %>%
#   ggplot() + 
#   geom_boxplot(aes(x=Enterotypes_id, y=dominance)) + 
#   ylab("alpha-diversity") +
#   xlab("DMM community types") +
#   cowplot::theme_cowplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) -> fig1B_dominance
# 
# 
# #fig1B
# fig1B_dominance



  

enterotypes %>%
  merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
  merge(enterotypes_phate_centroid, by="Enterotypes_id") %>%
  mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
  merge(genus_alpha_weight_curated, by.x="Enterotypes_id", by.y="variable") %>%
  mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
  ggplot() + 
  #geom_boxplot(aes(x=Enterotypes_id, y=shannon, fill=peripherality)) + 
  geom_boxplot(aes(x=Enterotypes_id, y=shannon)) + 
  #scale_fill_viridis_c(direction = -1, option="plasma") +
  ylab("alpha-diversity") +
  xlab("DMM community types") +
  cowplot::theme_cowplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) -> fig1A


```

```{r}
# 
# enterotypes %>%
#   merge(shannon_curated_v3, by.x="sample_name", by.y="row.names" ) %>%
#   mutate(Enterotypes_id = paste0("m",Enterotypes_id)) %>%
#   mutate(Enterotypes_id = Enterotypes_id %>% forcats::fct_reorder(shannon, median)) %>%
#   ggplot() + geom_bar(aes(x=Enterotypes_id)) +
#   xlab("DMM community types") +
#   ylab("number of\nsamples") +
#   cowplot::theme_cowplot() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) -> fig1C
#   
#   


```


### fig 1 ABC



```{r fig.height=8, fig.width=25}

library(patchwork)


#((fig1A|fig1B/fig1C) / ((fig1D + fig1E)/(fig1F +  fig1G)) + plot_layout(heights = c(0.75,1)) ) + plot_annotation(tag_levels = 'A')


(((fig1A/fig1B) + plot_layout(heights = c(0.25,1))) | (fig1C/fig1D) | (fig1E/fig1F) | (fig1G/fig1H)) + plot_annotation(tag_levels = 'A')


ggsave("figures/Figure1.pdf")

```

```{r fig.height=15, fig.width=15}

(
  (
    (((fig1A+ylab("alpha-\ndiversity"))/fig1B) + plot_layout(heights = c(0.25,1))) | 
    (fig1C/fig1D)
  ) / 
    (
      (fig1E/fig1F) | 
      (fig1G/(fig1H+cowplot::theme_cowplot()))
    )
  ) + plot_annotation(tag_levels = 'A')

ggsave("figures/Figure1_bis.pdf")


(
  (
    (((fig1A+ylab("alpha-\ndiversity"))/fig1B) + plot_layout(heights = c(0.25,1))) | 
      (fig1G/(fig1H+cowplot::theme_cowplot()))
    
  ) / 
    (
      (fig1E/fig1F) | 
      (fig1C/fig1D)
    )
  ) + plot_annotation(tag_levels = 'A')

ggsave("figures/Figure1_ter.pdf")


```


## link bifid cluster and age
```{r fig.height=5, fig.width=12}


sampleMetadata %>%
  mutate(age_category2 =
           case_when(
             age <= 3 ~ "00s",
             age <= 10 & age > 3 ~ "03+",
             age <=20 & age > 10 ~ "10s",
             age <=30 & age > 20 ~ "20s",
             age <=40 & age > 30 ~ "30s",
             age <=50 & age > 40 ~ "40s",
             age <=60 & age > 50 ~ "50s",
             age <=70 & age > 60 ~ "60s",
             age > 70 ~ "70+",
             TRUE ~ as.character(age),

           )) -> metadata

genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  merge(shannon_curated_v3, by.x="Row.names", by.y="row.names" ) %>%
  merge(metadata %>% select(sample_id,age_category2, age), by.x="Row.names", by.y="sample_id") %>%
  filter(!is.na(age)) %>%
  ggplot() + geom_point(aes(x=PHATE1, y=PHATE2, col=age+1), size=1, alpha=0.5) +
  scale_color_viridis_c("Age", option="plasma") + cowplot::theme_cowplot() + facet_wrap(~age_category2) -> figS7A



genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  merge(shannon_curated_v3, by.x="Row.names", by.y="row.names" ) %>%
  merge(metadata %>% select(sample_id,age_category2, age), by.x="Row.names", by.y="sample_id") %>%
  filter(!is.na(age)) %>%
  mutate(Enterotypes_id = paste("m",Enterotypes_id) %>% forcats::fct_reorder(age)) %>%
  ggplot() + geom_bar(aes(x=Enterotypes_id, fill=age_category2), position="fill") + 
  scale_fill_viridis_d("Age", option="plasma") +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(labels=scales::percent) +
  ylab("proportion of samples") +
  xlab("DMM community types") -> figS7B


genus_phate$embedding %>%
  merge(enterotypes, by.x="row.names", by.y="sample_name") %>%
  merge(shannon_curated_v3, by.x="Row.names", by.y="row.names" ) %>%
  merge(metadata %>% select(sample_id,age_category2, age), by.x="Row.names", by.y="sample_id") %>% 
  #filter(!is.na(age)) %>%
  mutate(Enterotypes_id = paste("m",Enterotypes_id) %>% forcats::fct_reorder(shannon)) %>%
  filter(age_category2=="00s" & !is.na(age)) %>%
  ggplot() + geom_bar(aes(x=Enterotypes_id, y = (..count..)/sum(..count..) )) + 
  #scale_fill_viridis_d("Age", option="plasma") +
  cowplot::theme_cowplot() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  scale_y_continuous(labels=scales::percent) +
  ylab("Infant samples (%)") +
  xlab("DMM community types") -> fig1C2


figS7A + figS7B + plot_annotation(tag_levels = "A")

ggsave("figures/figS7.pdf")

```

### figure 1 with infants
```{r fig.height=15, fig.width=15}

(((fig1A+ylab("alpha-\ndiversity"))/
  fig1B/
  fig1C2/
 (fig1H+cowplot::theme_cowplot()) + plot_layout(heights = c(0.50,2,0.50,1)) )|
  
  (
    
    (fig1E/fig1F/fig1C/fig1D)
    
  )) + plot_annotation(tag_levels = 'A')

ggsave("figures/Figure1_ter.pdf")

```


```{r fig.height=15, fig.width=15}

(((fig1A+ylab("alpha-\ndiversity"))/
  fig1B/
  fig1C2 + plot_layout(heights = c(1,2,1))
 |
  
  (
    
    (fig1F/fig1C/fig1D/fig1E/(fig1H+cowplot::theme_cowplot())))
    
  )) + plot_annotation(tag_levels = 'a')

ggsave("figures/Figure1_v4.pdf")


```


### genus painting

```{r fig.height=20, fig.width=30}

  
curated_v3_genus %>%
  tibble::column_to_rownames("Genus") %>%
  t() %>%
  as.data.frame() %>%
  tibble::rownames_to_column("sample_name") %>%
  select(sample_name, names(top_genus_mass_curated)) %>% #head %>%
  reshape2::melt(id.vars=c("sample_name","Bacteroides")) %>% #head %>%
  mutate(value=(value+1)/(Bacteroides+1)) %>% #head %>%
  merge(genus_phate$embedding %>% as.data.frame, by.x="sample_name", by.y="row.names") -> curated_v3_genus_PHATE_coord

curated_v3_genus_PHATE_coord %>%
  ggplot() + geom_point(aes(x=PHATE1,y=PHATE2,col=log10(value)), alpha=0.60) + scale_color_viridis_c() + facet_wrap(~variable)

ggsave("genus_phate_plot.pdf", h=20,w=30)


curated_v3_genus_PHATE_coord %>% 
merge(metadata %>% select(sample_id,age_category2, age),., by.x="sample_id", by.y="sample_name") %>%
  filter(!is.na(age)) %>%
  filter(age>=18) %>%
  ggplot() + geom_point(aes(x=PHATE1,y=PHATE2,col=log10(value)), alpha=0.60) + scale_color_viridis_c() + facet_wrap(~variable)

ggsave("genus_phate_plot_adults.pdf", h=20,w=30)
```
